### ISSUE TITLE
预训练时Loss一直不变

### ISSUE THREAD

**Issue by dayL-W** [ymcui/Chinese-LLaMA-Alpaca#441]
*提示：将[ ]中填入x，表示打对钩。提问时删除这行。只保留符合的选项。*

### 详细描述问题

准备了1个简单的中文语料文章，训练时发现Loss一直不变

训练脚本
>lr=2e-4
lora_rank=8
lora_alpha=32
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05
pretrained_model=/disk3/wuwen/stanford_alpaca/7B/alpaca-7b/
chinese_tokenizer_path=/disk3/wuwen/Chinese-LLaMA-Alpaca/scripts/chinese_sp.model
dataset_dir=/disk3/wuwen/Chinese-LLaMA-Alpaca/pdf2txt
data_cache=/disk3/wuwen/cache
per_device_train_batch_size=1
per_device_eval_batch_size=1
training_steps=100
gradient_accumulation_steps=1
output_dir=output_dir
#deepspeed_config_file=ds_zero2_no_offload.json
deepspeed_config_file=ds_zero3_offload.json
CUDA_VISIBLE_DEVICES=2
torchrun --nnodes 1 --nproc_per_node 1 run_clm_pt_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --data_cache_dir ${data_cache} \
    --validation_split_percentage 0.001 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --seed $RANDOM \
    --max_steps ${training_steps} \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.05 \
    --weight_decay 0.01 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 3 \
    --save_steps 50 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 1 \
    --block_size 128 \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --ddp_find_unused_parameters False \
    --fp16

ds_zero3_offload.json
>{
    "fp16": {
       "enabled": true,
       "auto_cast": false,
       "loss_scale": 0,
       "initial_scale_power": 16,
       "loss_scale_window": 1000,
       "hysteresis": 2,
       "min_loss_scale": 1
    },
    "zero_optimization": {
       "stage": 3,
       "offload_optimizer": {
          "device": "cpu",
          "pin_memory": true
       },
       "offload_param": {
          "device": "cpu",
          "pin_memory": true
       },
       "overlap_comm": true,
       "contiguous_gradients": true,
       "reduce_bucket_size": 205520896,
       "stage3_prefetch_bucket_size": 184968807,
       "stage3_param_persistence_threshold": 143360,
       "sub_group_size": 1e9,
       "stage3_max_live_parameters": 1e9,
       "stage3_max_reuse_distance": 1e9,
       "stage3_gather_16bit_weights_on_model_save": true
    },
    "steps_per_print": 100,
    "train_batch_size": 1,
    "train_micro_batch_size_per_gpu": 1,
    "gradient_accumulation_steps": 1,
    "wall_clock_breakdown": false,
    "optimizer": {
      "type": "Adam",
      "params": {
        "adam_w_mode": true,
        "lr": 2e-4,
        "betas": [
          0.9,
          0.999
        ],
        "eps": 1e-8,
        "weight_decay": 0.01
      }
    }
 }

### 运行截图或日志

![image](https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/29191972/2f1011cf-26d1-4297-8772-7b43ad720e30)

### 必查项目（前三项只保留你要问的）

- [x ] **基础模型**：LLaMA
- [x] **运行系统**：Ubuntu 
- [x ] **问题分类**：模型训练与精调 
- [ x] （必选）由于相关依赖频繁更新，请确保按照[Wiki](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki)中的相关步骤执行
- [ x] （必选）我已阅读[FAQ章节](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki/常见问题)并且已在Issue中对问题进行了搜索，没有找到相似问题和解决方案
- [x ] （必选）第三方插件问题：例如[llama.cpp](https://github.com/ggerganov/llama.cpp)、[text-generation-webui](https://github.com/oobabooga/text-generation-webui)、[LlamaChat](https://github.com/alexrozanski/LlamaChat)等，同时建议到对应的项目中查找解决方案

**Comment by ymcui** [comment]
```
chinese_tokenizer_path=/disk3/wuwen/Chinese-LLaMA-Alpaca/scripts/chinese_sp.model
```
这个传进去的词表文件不对啊，你要传lora压缩包里的tokenizer.model

**Comment by dayL-W** [comment]
> 

喔喔，我还以为这个是合并后的词表，我找找

**Comment by dayL-W** [comment]
> 

我使用merge_tokenizers.py脚本，将llama自带的token和chinese_sp.model合并在一起也可以吧

**Comment by ymcui** [comment]
这个目录里没有吗？你下载的Lora压缩包里就有tokenizer.model
```
pretrained_model=/disk3/wuwen/stanford_alpaca/7B/alpaca-7b/
```

**Comment by dayL-W** [comment]
> 这个目录里没有吗？你下载的Lora压缩包里就有tokenizer.model
> 
> ```
> pretrained_model=/disk3/wuwen/stanford_alpaca/7B/alpaca-7b/
> ```
> 
> 或者从huggingface单独下载tokenizer：
> 
> * alpaca: https://huggingface.co/ziqingyang/chinese-alpaca-lora-7b/tree/main
> * llama: https://huggingface.co/ziqingyang/chinese-llama-lora-7b/tree/main

包里有这个文件的，但是使用包里原本的还是作者提供的2个文件，Loss还是一直不变

**Comment by dayL-W** [comment]
> 这个目录里没有吗？你下载的Lora压缩包里就有tokenizer.model
> 
> ```
> pretrained_model=/disk3/wuwen/stanford_alpaca/7B/alpaca-7b/
> ```
> 
> 或者从huggingface单独下载tokenizer：
> 
> * alpaca: https://huggingface.co/ziqingyang/chinese-alpaca-lora-7b/tree/main
> * llama: https://huggingface.co/ziqingyang/chinese-llama-lora-7b/tree/main

我可能是没有做第一阶段Embeding预选了导致的这个问题？

**Comment by airaria** [comment]
权重是否有更新？可否比较一下训练结束后的模型某一层的权重和初始时模型某一层的权重？

**Comment by dayL-W** [comment]
> 

我打印了某层的参数，不是一模一样的，但是差别非常非常小。

训练前：
![image](https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/29191972/34e27945-855b-4112-a619-8260cf4daded)

训练后：
![image](https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/29191972/40289a87-cd50-48c3-906a-15801b5b5a7f)

训练过程Loss
![image](https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/29191972/4da56220-3cfb-4c35-b230-cc1524427d54)

训练日志LR显示是0，但实际上是2e-2，感觉日志展示的LR做四舍五入了，应该不是这个问题。
然后想到可能是 loss_scale=0 的问题（虽然等于0是自动缩放），设置了 loss_scale=65536，结果还是一样的

**Comment by dayL-W** [comment]
> 权重是否有更新？可否比较一下训练结束后的模型某一层的权重和初始时模型某一层的权重？

@airaria 我觉得可能是学习率的问题？我将学习率调的非常大5e-1，Loss立马下降到0，但这也是有问题的
![image](https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/29191972/f5b2d9b7-1c04-4679-a2f5-1e2b19c2232b)

**Comment by airaria** [comment]
2e-4是正常范围内的学习率，5e-1有些太大了。
只训了100步吗？脚本中的设置的100步仅是示例用，训久一点试试

**Comment by dayL-W** [comment]
> 2e-4是正常范围内的学习率，5e-1有些太大了。 只训了100步吗？脚本中的设置的100步仅是示例用，训久一点试试

@airaria 感谢回答，我重新使用了2e-4的学习率，batch_size=1，训练的语料是14个txt，预估10万个token左右，训练了3000多个step了（0.4epoch），但是loss还是一直为10.8203。那这样是不是说明，说明我准备的语料，这个模型都会，没有可提升空间？或者说训练的时间还是不够长？

**Comment by dayL-W** [comment]
> 

@airaria 
peft 库是按照提示来安装的指定版本。
因为在V100上训练的，使用zero 2报OOM，我看issues上，有人使用zero 3+V100跑成功了，就拿来试了试。有可能是这个问题

**Comment by airaria** [comment]
可以先不优化embed_tokens和lm_head并把长度调小试试能不能装下

**Comment by dayL-W** [comment]
> 可以先不优化embed_tokens和lm_head并把长度调小试试能不能装下

zero 2跑通了，但Loss还是一直不变，太难了
![image](https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/29191972/187f790f-0193-4e6b-bc16-85ff16daf270)



`lr=2e-4
lora_rank=8
lora_alpha=32
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.01

pretrained_model=/disk3/wuwen/stanford_alpaca/7B/alpaca-7b/
chinese_tokenizer_path=./merged_tokenizer_hf/tokenizer.model
dataset_dir=/disk3/wuwen/Chinese-LLaMA-Alpaca/pdf2txt
data_cache=/disk3/wuwen/cache
per_device_train_batch_size=1
per_device_eval_batch_size=1
training_steps=10000
gradient_accumulation_steps=1
output_dir=output_dir

export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda-12.0/targets/x86_64-linux/lib/

deepspeed_config_file=ds_zero2_no_offload.json
# deepspeed_config_file=ds_zero3_offload.json
CUDA_VISIBLE_DEVICES=2
torchrun --nnodes 1 --nproc_per_node 1 run_clm_pt_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --data_cache_dir ${data_cache} \
    --validation_split_percentage 0.001 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train True\
    --seed $RANDOM \
    --max_steps ${training_steps} \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.05 \
    --weight_decay 0.01 \
    --logging_strategy steps \
    --logging_steps 5 \
    --save_strategy steps \
    --save_total_limit 3 \
    --save_steps 500 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 1 \
    --block_size 256 \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --ddp_find_unused_parameters False \
    --fp16`

ds_zero2_no_offload.json

`{
    "fp16": {
        "enabled": true,
        "loss_scale": 0,
        "loss_scale_window": 100,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1e-10
    },

    "zero_optimization": {
        "stage": 2,
        "allgather_partitions": true,
        "allgather_bucket_size": 1e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 1e8,
        "contiguous_gradients": true,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
         },
         "offload_param": {
            "device": "cpu",
            "pin_memory": true
         }
    },

    "steps_per_print": 100,
    "train_batch_size": 1,
    "train_micro_batch_size_per_gpu": 1,
    "gradient_accumulation_steps": 1,
    "wall_clock_breakdown": false,
    "optimizer": {
      "type": "AdamW",
      "params": {
        "adam_w_mode": true,
        "lr": 2e-4,
        "betas": [
          0.9,
          0.999
        ],
        "eps": 1e-8,
        "weight_decay": 0.01
      }
    }
}
`

**Comment by gouqi666** [comment]
哈喽，解决了吗，我也遇到了类似的现象？
